### ISSUE TITLE
Combining unless-stopped and on-failure restart policies

### ISSUE THREAD

**Issue by bobvanderlinden** [docker/docker#17272]
With PR #15348 `--restart=unless-stopped` was added as an option. This is a great alternative for `--restart=always`, but it doesn't cover cases where you want `--restart=on-failure=...` as well as keeping the container stopped upon reboot when it was manually stopped earlier. Is there any option to do this already?

Side note: this is related to dokku where on-failure is now used by default. Each time an application is published by dokku a new container is created and the old container is stopped. However, upon rebooting the system, all old containers will start again too. Usually this means an old version of an application is listening on the published TCP port instead of the latest version. See #1505.

**Comment by tedyoung** [comment]
For similar reasons, it'd be great to have a combination of these two options. Is this possible?

**Comment by thaJeztah** [comment]
There's currently a PR that allows updating the restart-policy of an existing container; https://github.com/docker/docker/pull/19116

I think that would give more flexibility, as it will allow to remove the restart policy of a container if needed (i.e. dokku would be able to stop the container, and remove the restart policy)

wdyt?

**Comment by gdm85** [comment]
Also related: #8821

**Comment by francislavoie** [comment]
I have a similar issue as I described here: https://github.com/moby/moby/issues/8821#issuecomment-319308454

I want an `unless-stopped` policy that doesn't restart on errors (or configurable max-tries on errors).

**Comment by bobvanderlinden** [comment]
The option `always-stopped` is [documented](https://docs.docker.com/engine/reference/run/) as:

> Always restart the container regardless of the exit status, but do not start it on daemon startup if the container has been put to a stopped state before.

Whereas `on-failure` is documented as:

> Restart only if the container exits with a non-zero exit status. Optionally, limit the number of restart retries the Docker daemon attempts.

It seems that there isn't an option to have a container restarted upon failure, but also let it be in a stopped state once it was stopped manually.

**Comment by francislavoie** [comment]
@bobvanderlinden no, that's exactly what `unless-stopped` does and describes. "Always restart the container **regardless of the exit status**"

**Comment by bobvanderlinden** [comment]
What I'm looking for is one that restarts on failure (so it *should* restart upon a non-zero exit code), but should *not* restart when it is in stopped state. That means it also shouldn't restart when the system boots up while the container was in stopped state before the reboot.

Basically:

<table>
<tr><th>Restart option<</th><th>Failure exit code in running state</th><th>Non-failure exit in running state</th><th>Stopped state</th></tr>
<tr><td>`unless-stopped`</td><td>✔</td><td>✔</td><td>✗</td></tr>
<tr><td>`on-failure`</td><td>✔</td><td>✗</td><td>✔</td></tr>
<tr><td>`always`</td><td>✔</td><td>✔</td><td>✔</td></tr>
<tr><td><i>My wish</i></td><td>✔</td><td>✗</td><td>✗</td></tr>
<tr><td><i>Your wish</i></td><td>✗</td><td>✔</td><td>✗</td></tr>
</table>

I think *My wish* makes sense for a LOT of use-cases. Usually you don't want your container to be automatically restarted when you have manually stopped it. It also causes problems for Dokku and similar container-managing services.

To be honest, I think splitting the `--restart` option into `--restart-on-exit=(always|failure|success)` and `--restart=(unless-stopped|always)` options would avoid naming all the different combinations people will come up with, but still covers all cases.

**Comment by francislavoie** [comment]
Nice, that's pretty comprehensive. Thanks for clarifying

**Comment by francislavoie** [comment]
@thaJeztah Can we get a comment on this? I'm being forced to do a stupid hack to get around the lack of a proper restart policy for my situation. See https://github.com/abiosoft/caddy-docker/issues/65

**Comment by gugrim** [comment]
TLDR:
Add a "Maximum Retry Count" option to "unless-stopped".

Explanation:
Recently, my volume disk for my mssql container (SQL Server) filled up due to core dumps followed by restart, in a loop. It can't handle kernel 6.7. I had the restart policy "unless-stopped" since I want the mssql container to start after a host reboot or docker service restart. IMO, the "on-failure" policy is basically broken. I guess it can be used in some situations, since it exists, but I have not seen a use for it without some outside help such as starting containers in a boot script or service that depends on docker. But the docker docs says "Don't combine Docker restart policies with host-level process managers, as this creates conflicts.".

Changing the "on-failure" policy to start such containers on docker start would be my preferred solution, that is the way I think it should work, but I guess that ship has sailed.

Perhaps the simplest solution would be to add a "Maximum Retry Count" to "unless-stopped". Then docker would start such containes when it starts up, unless they have reached their retry count, in which case they need to be started manually.

**Comment by soleous** [comment]
I am being hit by this for a project I'm working on.  Where I need the container to fail (stop) on error due to resources, but start on restart.  

I am strugging to find a solution where I can force the container to stop (while using unless-stopped) from within the container as a workaround.  I know this is a long standing issue, but has anyone got a viable workaround for this.
